cli/internal/codegen/golang.go

```diff
@@ -0,0 +1,821 @@
+package codegen
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/cockroachdb/errors"
+	. "github.com/dave/jennifer/jen"
+
+	"encr.dev/cli/internal/version"
+	meta "encr.dev/proto/encore/parser/meta/v1"
+	schema "encr.dev/proto/encore/parser/schema/v1"
+)
+
+type golang struct {
+	md *meta.Data
+}
+
+func (g *golang) Generate(buf *bytes.Buffer, appSlug string, md *meta.Data) (err error) {
+	g.md = md
+
+	namedTypes := getNamedTypes(md)
+
+	// Create a new client file
+	file := NewFile("client")
+
+	// Generate the parent Client struct
+	g.generateClient(file, appSlug, md.Svcs)
+
+	// Generate the types and service client structs
+	for _, service := range md.Svcs {
+		g.generateTypeDefinitions(file, namedTypes.Decls(service.Name))
+
+		if hasPublicRPC(service) {
+			g.generateServiceClient(file, service)
+		}
+	}
+
+	// Generate the base client
+	g.generateBaseClient(file)
+
+	// Finally, render the client
+	if err := file.Render(buf); err != nil {
+		return errors.Wrap(err, "unable to generate go client")
+	}
+
+	return nil
+}
+
+func (g *golang) cleanServiceName(service *meta.Service) string {
+	return strings.Title(strings.ToLower(service.Name))
+}
+
+// generateClient creates the Client struct, Option type and New function
+func (g *golang) generateClient(file *File, appSlug string, services []*meta.Service) {
+	// List all services which have public RPCs or types we need
+	fieldDef := make([]Code, 0, len(services))
+	fieldInit := make(Dict)
+	for _, service := range services {
+		if !hasPublicRPC(service) {
+			continue
+		}
+
+		name := g.cleanServiceName(service)
+
+		fieldDef = append(fieldDef,
+			Id(name).Id(fmt.Sprintf("%sClient", name)),
+		)
+
+		fieldInit[Id(name)] = Op("&").Id(fmt.Sprintf("%sClient", strings.ToLower(name))).Values(Id("base"))
+	}
+
+	// The client struct
+	file.Commentf("Client is a generated client for the Encore application %s.", appSlug)
+	file.Add(
+		Type().Id("Client").Struct(fieldDef...),
+		Line(),
+		Line(),
+	)
+
+	file.Comment("BaseURL is what URL your encore application is being hosted at")
+	file.Type().Id("BaseURL").String()
+	file.Line()
+
+	file.Const().Id("LocalDevelopment").Id("BaseURL").Op("=").Lit("http://localhost:4000")
+	file.Line()
+
+	file.Comment("Environment can be used to specify an environment by name, rather than using LocalDevelopment.")
+	file.Comment("")
+	file.Comment("This function doesn't check if the environment exists first, so it can cause runtime errors if")
+	file.Comment("you pass in an invalid environment name")
```

Comment:
I think the comment about runtime errors is a bit verbose and unnecessary:
```suggestion
	file.Comment("Environment returns a BaseURL for calling the cloud environment with the given name.")
```